### DISCUSSION TITLE
How to add WebApi Versioning?

### DISCUSSION THREAD

**Discussion by eugene-latarevich** [OrchardCMS/OrchardCore#15966] [Q&A]
Could somebody explain me how to add WebApi Versioning for an Api Module?
I tried to add AddApiVersioning extension to services and UseApiVersioning for IApplicationBuilder but it doesn't work.

Route atribute for api controller method looks like [Route("api/v{version:apiVersion}/get")] for test purposes.

Thank you in advance.

**Comment by jtkech** [comment]
Maybe the problem is that you are doing things through the application startup, if so, because everything is run in a tenant context composed of modules, you would have to use the following helper that allows to configure things from the application but as it is done through a module.

            AddOrchardCms(services, null)
                .AddOrchardCore()
                .Configure(app =>
                {
                    // use the passed tenant app builder
                });

            return services;

With the dev branch `AddOrchardCms()` returns an `OrchardCoreBuilder`, so you can do directly

            AddOrchardCms(services, null)
                .Configure(app =>
                {
                    // use the passed tenant app builder
                });

            return services;

**Comment by eugene-latarevich** [comment]
I've added to source
public static IServiceCollection AddOrchardCms(
            IServiceCollection services, 
            Action<OrchardCoreBuilder> configure
        )
        {
            var builder = services.AddOrchardCore()

                .AddCommands()

                .AddMvc()
                .AddSetupFeatures("OrchardCore.Setup")

                .AddDataAccess()
                .AddDataStorage()
                .AddBackgroundService()

                .AddTheming()
                .AddLiquidViews()
                .AddCaching()
                .Configure(app => 
                {
                    app.UseApiVersioning();
                });

            // OrchardCoreBuilder is not available in OrchardCore.ResourceManagement as it has to
            // remain independent from OrchardCore.
            builder.ConfigureServices(s =>
            {
                s.AddResourceManagement();

                s.AddTagHelpers<LinkTagHelper>();
                s.AddTagHelpers<MetaTagHelper>();
                s.AddTagHelpers<ResourcesTagHelper>();
                s.AddTagHelpers<ScriptTagHelper>();
                s.AddTagHelpers<StyleTagHelper>();

                s.AddApiVersioning();
            });

            configure?.Invoke(builder);

            return services;
        }
But I have the error:
An unhandled exception occurred while processing the request.
NullReferenceException: Object reference not set to an instance of an object.
Microsoft.AspNetCore.Mvc.Routing.ApiVersionMatcherPolicy.EvaluateApiVersion(CandidateSet candidates, ApiVersion apiVersion)

Stack Query Cookies Headers Routing
NullReferenceException: Object reference not set to an instance of an object.
Microsoft.AspNetCore.Mvc.Routing.ApiVersionMatcherPolicy.EvaluateApiVersion(CandidateSet candidates, ApiVersion apiVersion)
Microsoft.AspNetCore.Mvc.Routing.ApiVersionMatcherPolicy.ApplyAsync(HttpContext httpContext, CandidateSet candidates)
Microsoft.AspNetCore.Routing.Matching.DfaMatcher.SelectEndpointWithPoliciesAsync(HttpContext httpContext, IEndpointSelectorPolicy[] policies, CandidateSet candidateSet)
Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.<Invoke>g__AwaitMatch|8_1(EndpointRoutingMiddleware middleware, HttpContext httpContext, Task matchTask)
OrchardCore.Diagnostics.DiagnosticsStartupFilter+<>c__DisplayClass3_0+<<Configure>b__1>d.MoveNext() in DiagnosticsStartupFilter.cs
Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
OrchardCore.ContentPreview.PreviewStartupFilter+<>c+<<Configure>b__1_1>d.MoveNext() in PreviewStartupFilter.cs
Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func<ShellScope, Task> execute) in ShellScope.cs
OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in ModularTenantContainerMiddleware.cs
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)

Where am I wrong?

**Comment by jtkech** [comment]
Looks like the same issue we had with the mvc `PageLoaderMatcherPolicy` that was not doing the following check and that was failing on an NRE.

        ...
        for (var i = 0; i < candidates.Count; i++)
        {
            // Additional checking.
            if (!candidates.IsValidCandidate(i))
            {
                continue;
            }
        ...

So before they fix it, we had to copy paste the related file to do a custom implementation with the above checking and then replace the original implementation in the DI.

See [this mvc issue](https://github.com/dotnet/aspnetcore/issues/13996) for more details and where i explained how to replace the original implementation in the DI.

**Comment by sebastienros** [comment]
@jtkech did you report the issue?

**Comment by jtkech** [comment]
@sebastienros no but i will do

**Comment by jtkech** [comment]
Done https://github.com/microsoft/aspnet-api-versioning/issues/600
